Enable 2d + AC torch.compile

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112536

This PR enables AC + torch.compile to work with FSDP + TP, the fix to
high order op path is that we need to check both tensor and tensor
subclass bases to make sourceless builder

NOTE: selective AC + 2D is still not working, need to fix this
separately

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

wanchaol committed:
Enable 2d + AC torch.compile

This PR enables AC + torch.compile to work with FSDP + TP, the fix to
high order op path is that we need to check both tensor and tensor
subclass bases to make sourceless builder

NOTE: selective AC + 2D is still not working, need to fix this
separately

[ghstack-poisoned]

wanchaol committed:
Update on "Enable 2d + AC torch.compile"


This PR enables AC + torch.compile to work with FSDP + TP, the fix to
high order op path is that we need to check both tensor and tensor
subclass bases to make sourceless builder

NOTE: selective AC + 2D is still not working, need to fix this
separately

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

fduwjj commented on test/distributed/_tensor/test_dtensor_compile.py:
```diff
@@ -270,6 +281,49 @@ def test_2d_fsdp_tp_compile(self):
+            "mlp_1.net2": RowwiseParallel(),
+        }
+        tp_model = parallelize_module(model, mesh_2d["tp"], parallelize_plan)
+        tp_model = checkpoint_wrapper(
+            tp_model,
+            checkpoint_impl=CheckpointImpl.NO_REENTRANT,
+            checkpoint_fn=checkpoint,
+            use_reentrant=False,
+        )
```
Just one more point here, with TP AC, users usually use `input_reshard` together.

wanchaol replied:
hmmm I don't understand in what case user would use `input_reshard`, we don't use it at least

wanchaol replied:
double checked on this `input_reshard`, you mean for TP only (not SP) this would reduce the memory usage when combining with FSDP checkpoint wrapper? wondering why we didn't apply this to our models when applying TP, we should probably add it. Is it because baseline don't have this feature?

zou3519 commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1095,7 +1095,9 @@ def call_getattr(
-                        bases[0] is object or bases[0] is torch._C.TensorBase
+                        bases[0] is object
+                        or bases[0] is torch._C.TensorBase
+                        or bases[0] is torch.Tensor
```
can you add a more direct test for this somewhere (in maybe test/dynamo/test_misc.py?) Brian told me this was necessary for the case where we construct a Tensor subclass and attempt to trace over a function that does tree_flatten on a container holding the subclass.

wanchaol replied:
Sure, let me see add a test case there

wanchaol replied:
do you think this is specifically related to high order ops so I need to add tests to test_high_order_ops.py or it's a general problem where pytree flatten a tensor subclass?

bdhirsh replied:
I don't think that this is related to higher order ops - we only ran into issues with the higher order op because this  was causing a graph break inside the body of our higher order op.

So one test I think we can write is just to `torch.compile(full_graph=True)` some code that tries to pytree a tensor subclass:
```
@torch.compile(backend="aot_eager", fullgraph=True)
def f(subclass):
    nested_list = [[subclass]]
    flattened_list, _ = pytree.tree_flatten(nested_list)
    return flattened_list[0]
```

wanchaol replied:
I wrote a test case like this but it does not seems fail even when we don't have this changes, weird.

```
    def test_tracing_py_tree_tensor_subclass(self):
        import torch.utils._pytree as pytree
        from torch.testing._internal.two_tensor import TwoTensor

        def fn(xs):
            nested_xs = [[xs]]
            flat_xs, spec = pytree.tree_flatten(xs)
            return flat_xs[0].clone()

        xs = TwoTensor(torch.ones(2, 2), torch.ones(2, 2))

        counter = CompileCounter()
        torch._dynamo.optimize(counter, nopython=True)(fn)(xs)
        self.assertEqual(counter.frame_count, 1)
        self.assertEqual(counter.op_count, 1)
```

bdhirsh replied:
oh hmm, it looks like this ends up working out because we hit this line, and the input subclass has a source in dynamo: https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/builtin.py#L1077

Hmm... I wonder how we can trigger the path where our subclass doesn't have a source (which is apparently what gets triggered in the checkpointing case

bdhirsh replied:
I'm not actually sure how to trigger a situation where dynamo doesn't have a source for a given tensor (maybe Voz/Yanbo/someone from dynamo can help).

But otherwise - do you think you can try adding a test that uses a subclass + activation checkpointing, and see if that can give us a repro?

zou3519 replied:
> Hmm... I wonder how we can trigger the path where our subclass doesn't have a source (which is apparently what gets triggered in the checkpointing case

Does it work if you construct a subclass inside of the function being torch.compile'd, then put that into a list, and attempt to tree_flatten it?

wanchaol replied:
Hmmm I don't think we support constructing a subclass inside a torch.compiled function, right now it would graph break when those case happened iiuc

wanchaol replied:
I'll try out next put subclass + AC together and see if that repros

wanchaol replied:
added AC together and it works now :) Thanks for suggesting!

wz337 commented on test/distributed/_tensor/test_dtensor_compile.py:
```diff
@@ -270,6 +281,49 @@ def test_2d_fsdp_tp_compile(self):
+        tp_degree = self.world_size // dp_degree
+        model = SimpleModel(self.device_type)
+        model_copy = copy.deepcopy(model)
+        enable_2d_with_fsdp()
```
We no longer need `enable_2d_with_fsdp` anymore with the device_mesh approach.

wanchaol replied:
sg, I'll delete it

wanchaol committed:
Update on "Enable 2d + AC torch.compile"


This PR enables AC + torch.compile to work with FSDP + TP, the fix to
high order op path is that we need to check both tensor and tensor
subclass bases to make sourceless builder

NOTE: selective AC + 2D is still not working, need to fix this
separately

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge -i "failures not related"

wanchaol commented:
@pytorchbot merge -i
